### DISCUSSION TITLE
Tutorial on how to create a package

### DISCUSSION THREAD

**Discussion by carlmathisen** [adonisjs/core#3492] [General]
Hi guys,

Thank you for a really great framework @thetutlage, @RomainLanz and team.

I've been working on a few packages recently internally for my work, but also a few that I want to open source. I've poked around in a few well written packages, such as the `@eidellev/inertia-adonisjs` package and others to try to figure out how to optimally setup a package the Adonis way.

I'm finishing up the packages now, but they initially didn't have:
* Their own IoC declarations
* Config and setup files in the standard spots (`contracts`, `start`, `config` - what goes where)
* Configuration steps with instructions (`node ace configure`)
* Japa tests (less important which test framework is being used though, but I migrated to it since it's the official one)

It would be useful if there was a tutorial or a boilerplate repo for package authors to get started. I think it would be great for Adonis adoption if it was easier to get started with creating packages. The quality of packages will be higher if we had an official take on how to create them.

**Comment by skrenek** [comment]
This has been discussed off and on in the community.  This guide you mention is on the road map for creation.  If you haven't already I'd recommend voting for it on the [trello roadmap](https://trello.com/b/3klaHbfP/adonisjs-roadmap).  In the meantime, you've already done what the rest of us who have wanted to create our own packages have; reverse engineered. 😄

**Comment by guoyunhe** [comment]
It has been a year. Still no guide on creating packages...

**Comment by McSneaky** [comment]
Might be worth to go thro that: https://mcsneaky.ap3k.pro/posts/adonis-v5-provider/#how-to-turn-it-into-external-sharable-provider

There's also whole series by Chimezie on Adonis Mastery https://adonismastery.com/courses/adonisjs-package-development

Also guide on Adocasts https://adocasts.com/lessons/how-to-make-a-simple-adonisjs-package

I know Virk is doing a lot of internal cleanup to make package development a lot easier and to get rid of some odd hacks. Most likely what that's ready there will be some official docs / videos / podcasts about it. That will still take some time tho

**Comment by seekcx** [comment]
I'm trying to gather existing material to develop a package and it seems to be difficult. I also looked at the source code of some packages, but I am not sure whether his solution is adonis'. If there is a systematic document or starter project, I think the ecology of adonis will develop faster. What stops me at present is that I don't know how to start.

Hope the official development team can pay attention to this issue.

**Reply by McSneaky** [reply]
Jup, official doc for it is in development :)

**Reply by rbartholomay** [reply]
any news on this?

**Comment by rbartholomay** [comment]
Hi!

Have a look at:  

https://v6-migration.adonisjs.com/guides/package-migration and https://github.com/adonisjs/pkg-starter-kit!

And for Adonis5 there is an video: https://adocasts.com/lessons/how-to-make-a-simple-adonisjs-package


Bye, René
